---
title: "Estimating directional epistasis"
author: "Salomé Bourg"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Estimating directional epistasis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

This vignette shows how to estimate the directional epistasis in a morphological trait between two diverged populations using `linecross::lcross`. The function `lcross` can fit 15 models with different reference genotypes on line-cross data in order to study directional epistasis. 

A line-cross consists of a series of crosses between two populations that have diverged via artificial or natural selection, or between inbred lines, to produce a set of standard populations, referred to as derivatives, with specific genetic properties that can be computed from quantitative genetic theory. The two parental populations, P1 and P2, assumed to be in Hardy-Weinberg equilibrium, are crossed to produce a first-generation hybrid, the F1. A second series of crosses between F1 individuals produces second-generation hybrids, the F2 population. Backcrosses B1 and B2 are produced by backcrossing the F1 with the two parental populations P1 and P2.


```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(linecross)
```
# The `lcross` function

## The data format
Full reciprocal crosses imply a lot of different crosses, as shown in the figure below. The function `lcross` requires a specific data.frame format for the argument `data` with three columns (derivatives, means and standard errors) and 16 rows. Each row corresponds to one specific derivative population (a blue circle in the figure).

```{r, out.width = '100%', echo = FALSE}
## fig.height = 2, fig.width = 5, fig.align = "center"
knitr::include_graphics("img/reciprocal_explanation.png")
```

Your data have to follow the same format. If the full reciprocal crosses have not been done, the means and standard error of the missing derivative populations have to be NA, like in the example below. Reciprocal cross are often unknown or different reciprocal crosses are pooled, for example the if the data has only one F2 containing all types of reciprocal crosses. In this situation, it does not matter which of the different F2 entries that have NA, but one cannot fit models that relies on reciprocal information to such data. Note that there are no models requiering resprocal information that is implemented in this package. 


```{r}
data_test=data.frame(
  pop  = c("P1", "P2", "F1", "F2", "B1", "B2", "P1xF2", "F1xB2"),
  mean = c(0.5,  2.0,  1.5,   1.0, 1.0,   1.7,  1.0,    1.5), 
  se   = rep(0.1, 8))

data_test
```

## Choice of model
The multilinear model is useful for studying directional epistasis in the divergence of one population from another.

## Choice of reference
Since we want to study directional epistasis between two diverged populations, for this question it makes most sense to have the ancestral population as P1 and use this as the reference. We defined the reference population P1 as the parental population with the smallest trait mean. By doing so, the direction of epistasis corresponds to the effect of allele substitution when selection increases the trait size.

```{r}
mod.multilinear=lcross(model="multilinear", ref="P1", data=data_test, maxeval=3)
cbind(mod.multilinear$parameters, mod.multilinear$parameters.se)
```
## Computing epistasis factor
Hansen and Wagner (2001) defined a dimensionless epistasis factor that describes the rescaling of genetic effects during subsequent substitutions. In the context of line crosses this corresponds to the factor the predicted value obtained with the model containing only the reference effects (i.e. without epistasis) should be multiplied to obtain the predicted value with epistasis (Bolstad et al. 2023):

```{r, eval = FALSE}

epfac(mod.multilinear)


# episfac.p1.p2.multi= 1 + (mod.multilinear$parameters["eps"]/2)*mod.multilinear$parameters["Y2"]
# names(episfac.p1.p2.multi) <- "episfac"
# episfac.p1.p2.multi
```

The epistasis factor is 0<ef<1 corresponding to negative epistasis. This epistasis factor indicates that the P2 phenotype is only 35% of the expected phenotype under an additive model. In other words, epistasis generates a 65% (1 – 0.35) decrease of the effect of allele substitutions compared with the additive model when the trait mean (or the genotype) shifts from P1 to P2. This result also indicates that additive variance in the P2 population is expected to be ca. 12% (0.35^2) of the additive variance observed in the P1 population for this given trait. 

# The `plot_linecross` function

The axes are genotypes (proportion of P1 alleles, denoted 1-S), heterozygosity (the probability that an individual has one P1 and one P2 allele at a locus, denoted H) and phenotypes. The two variables S and H define a two-dimensional genetic-content space, with P1, F1 and P2 at the edges. Using these variables in a regression with the phenotype as the response variable, this triangle becomes a surface where the height gives the genotypic value (i.e. the trait value of each genotype). This surface is a description of a two-dimensional genotype-phenotype map representing the composite (genome wide sum) genetic effects.

```{r,fig.height = 4.2, fig.width = 7, fig.align = "center"}
plot_linecross(output.model=mod.multilinear, data=data_test, theta=-30, phi=20, margins=c(2,2,2,2), col.triangle="gray80")
```

The arguments `theta` and `phi` allow for twisting the orientation of the 3D plot (same as in `persp`) and `col.triangle` determines the colors of the surfance triangle formed by the line-cross analysis. Black dots represent the observed mean for the populations and derivatives and grey dots are the predicted values.

<br>

The GP-map exhibits strong negative epistasis from P1 to P2 (i.e., the surface is concave in the S-direction). 


# References
Bolstad GH, Bourg S, Griffin DV, Pélabon C & Hansen TF. 202?. Quantifying genome-wide functional epistasis by line-cross analysis

Bourg S, Bolstad GH, Griffin DV, Pélabon C & Hansen TF. 202?. Directional epistasis is common in morphological traits.

Hansen, T. F., & Wagner, G. P. (2001). Modeling genetic architecture: a multilinear theory of gene interaction. Theoretical population biology, 59(1), 61-86.
